src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -284,16 +284,15 @@ public static void AddToStringBuilder(ReadOnlyMemory<char> memory, StringBuilder
                 sb.Append(memory);
         }
 
-        public static void AddLowerCaseToStringBuilder(ReadOnlyMemory<char> memory, StringBuilder sb)
+        public static void AddLowerCaseToStringBuilder(ReadOnlySpan<char> span, StringBuilder sb)
```

Comment:
Hmm, instead of this custom implementation that traverses the span a character at a time and does the check if ToLower changed it, maybe use the span.ToLowerInvariant(span) to first lower case the entire span, and then just call Append on it directly?

Granted, it would require some unsafe code to change the ReadOnlySpan to a Span, since the destination argument needs to be writable.

Is that feasible?

Reply:
You would also need to allocate new memory (the receiving span) that way.  Here we are trying to append the `ToLowerInvariant` string into an existing `StringBuilder`. If we took that approach, we would need to allocate some memory to write the `ToLowerInvariant` string into, and then append that to `StringBuilder` - i.e. 2 copies.

IMO it would be better if we had a `span.ToLowerInvariant(StringBuilder)` overload in the framework.  Or some way of getting a `Span` from `StringBuilder` that was of a sufficient size.

Reply:
> You would also need to allocate new memory (the receiving span) that way.

I was thinking of using the input as the destination. Something like this:
```C#
ReadOnlySpan<char> span = "aAbBcCedbaG";
Span<char> destination = MemoryMarshal.CreateSpan(ref MemoryMarshal.GetReference(span), span.Length);
span.ToLowerInvariant(destination);
sb.Append(span); // or destination
```

Reply:
It’s not acceptable to change the input.

Reply:
> It’s not acceptable to change the input.

I see. Then, nvm.